bootloader/src/rsdp-finder.c

```diff
@@ -0,0 +1,30 @@
+#include "rsdp-finder.h"
+
+// I could not find any functions comparing two EFI_GUIDs.
+
+// macro of GUID must be put in memory for comparisons
+const EFI_GUID acpi_guid_bytes = ACPI_20_TABLE_GUID;
+const uint8_t* acpi_guid = (uint8_t*)&acpi_guid_bytes;
+
+EFI_PHYSICAL_ADDRESS find_rsdp(EFI_SYSTEM_TABLE* st) {
+
+    // iterate config table
+    for (UINTN i = 0; i < st->NumberOfTableEntries; i++) {
+        EFI_CONFIGURATION_TABLE* table = st->ConfigurationTable + i;
+
+        const uint8_t* guid = (uint8_t*)&table->VendorGuid;
+
+        { // memcmp
+            int j = 0;
+            while (j < 16 && guid[j] == acpi_guid[j]) j++;
+
+            // if entire guid matched
+            if (j == 16) {
+                return (EFI_PHYSICAL_ADDRESS)table->VendorTable;
+            }
+        }
+    }
+
+    // no rsdp could be found for ACPI 2.0
+    return 0;
```

Comment:
Maybe return EFI_NOT_FOUND instead of zero and have an out parameter for the address because zero is a valid address. Then you could return EFI_SUCCESS on success. Take a look at load_kernel function.